### ISSUE TITLE
EMC Link - 1.18.2

### ISSUE THREAD

**Issue by DinnerBeef** [FTBTeam/ProjectEX#124]
EMC links only show 1 in the system not the max amount

**Comment by desht** [comment]
Yeah, was wondering about the best approach here.  Ideally it should show just the number that could be pulled based on the player's EMC, but that might get laggy (needing to recalculate amounts all the time).

Or it could always show the max stack size, but that could also be a lie if the player doesn't have enough EMC to pull a stack...

**Comment by DinnerBeef** [comment]
Main thing is it need to be high enough for it to be usable in auto crafting. Crafting can sometimes need 10s of thousands of one type of item, if not 100 of thousands. A stack size won't even near be enough for it to show in the system.

**Comment by MaxNeedsSnacks** [comment]
> Yeah, was wondering about the best approach here. Ideally it should show just the number that could be pulled based on the player's EMC, but that might get laggy (needing to recalculate amounts all the time).
> 
> Or it could always show the max stack size, but that could also be a lie if the player doesn't have enough EMC to pull a stack...

This is definitely a bit of an issue since I remember EMC Links causing a fair bit of tick lag in NTC2 due to having to recalculate the counts, so while it *is* the most "correct" approach here the refresh rate would definitely be a bit of an issue.

**Comment by DinnerBeef** [comment]
They only way I can think of getting around recalculating all the time is having the EMC link act more like a crafter so if there is not enough it can "craft" more. However, you would still want to break the functions It has right now, being able to pull items out of it.

**Comment by desht** [comment]
There are ways of mitigating performance issues here with some smart caching of the block's knowledge capability and the player it's connected to (although great care will be needed to ensure cache validity).

Ultimately though, if you put a Storage Bus/External Storage/etc. on one of these blocks, it will get polled for items on a very regular basis.  So a TPS hit is inevitable, but I'll do what I can to minimize the hit.
